// @ngxpert/hot-toast/fesm2022/ngxpert-hot-toast.mjs
import { DOCUMENT, isPlatformServer } from '@angular/common';
import * as i0 from '@angular/core';
import { input, ChangeDetectionStrategy, Component, output, signal, inject, Injector, Renderer2, NgZone, ChangeDetectorRef, ViewChild, Input, ElementRef, afterNextRender, isDevMode, ViewChildren, InjectionToken, PLATFORM_ID, Injectable, makeEnvironmentProviders } from '@angular/core';
import { DynamicViewDirective, isTemplateRef, isComponent, ViewService } from '@ngneat/overview';
import { Subject, race, defer, Subscription, throwError } from 'rxjs';
import { map, filter, tap, catchError } from 'rxjs/operators';
import { HttpErrorResponse } from '@angular/common/http';

const HOT_TOAST_DEFAULT_TIMEOUTS = {
    blank: 4000,
    error: 4000,
    success: 4000,
    loading: 30000,
    warning: 4000,
    info: 4000,
};
const EXIT_ANIMATION_DURATION = 800;
const ENTER_ANIMATION_DURATION = 350;
const HOT_TOAST_MARGIN = 8;
const HOT_TOAST_DEPTH_SCALE = 0.05;
const HOT_TOAST_DEPTH_SCALE_ADD = 1;
/**
 * Per-status defaults applied when using `fromForm`.
 * Centralised here so future tweaks only need one change.
 */
const HOT_TOAST_FORM_STATUS_DEFAULTS = {
    VALID: { type: 'success' },
    INVALID: { type: 'error' },
    PENDING: { type: 'loading' },
    DISABLED: { type: 'blank' },
};

class HotToastRef {
    constructor(toast) {
        this.toast = toast;
        this.groupRefs = [];
        this.groupExpanded = false;
        /** Subject for notifying the user that the toast has been closed. */
        this._onClosed = new Subject();
        /** Subject for notifying the user that the toast has been closed. */
        this._onGroupToggle = new Subject();
    }
    set data(data) {
        this.toast.data = data;
    }
    get data() {
        return this.toast.data;
    }
    set dispose(value) {
        this._dispose = value;
    }
    getToast() {
        return this.toast;
    }
    /**
     * Used for internal purpose
     * Attach ToastRef to container
     */
    appendTo(container, skipAttachToParent) {
        const { dispose, updateMessage, updateToast, afterClosed, afterGroupToggled, afterGroupRefsAttached } = container.addToast(this, skipAttachToParent);
        this.dispose = dispose;
        this.updateMessage = updateMessage;
        this.updateToast = updateToast;
        this.afterClosed = race(this._onClosed.asObservable(), afterClosed);
        this.afterGroupToggled = race(this._onGroupToggle.asObservable(), afterGroupToggled);
        this.afterGroupRefsAttached = afterGroupRefsAttached;
        return this;
    }
    /**
     * Closes the toast
     *
     * @param [closeData={ dismissedByAction: false }] -
     * Make sure to pass { dismissedByAction: true } when closing from template
     * @memberof HotToastRef
     */
    close(closeData = { dismissedByAction: false }) {
        this.groupRefs.forEach((ref) => ref.close());
        this._dispose();
        this._onClosed.next({ dismissedByAction: closeData.dismissedByAction, id: this.toast.id });
        this._onClosed.complete();
    }
    toggleGroup(eventData = { byAction: false }) {
        this.groupExpanded = !this.groupExpanded;
        this._onGroupToggle.next({
            byAction: eventData.byAction,
            id: this.toast.id,
            event: this.groupExpanded ? 'expand' : 'collapse',
        });
    }
    show() {
        this.toast.visible = true;
    }
}

const animate = (renderer, element, animation) => {
    renderer.setStyle(element, 'animation', animation);
};

class LoaderComponent {
    constructor() {
        this.theme = input(...(ngDevMode ? [undefined, { debugName: "theme" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: LoaderComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.2.9", type: LoaderComponent, isStandalone: true, selector: "hot-toast-loader", inputs: { theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "<div\n  class=\"hot-toast-loader-icon\"\n  [style.border-color]=\"theme()?.primary\"\n  [style.border-right-color]=\"theme()?.secondary\"\n></div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: LoaderComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-loader', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div\n  class=\"hot-toast-loader-icon\"\n  [style.border-color]=\"theme()?.primary\"\n  [style.border-right-color]=\"theme()?.secondary\"\n></div>\n" }]
        }], propDecorators: { theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }] } });

class ErrorComponent {
    constructor() {
        this.theme = input(...(ngDevMode ? [undefined, { debugName: "theme" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: ErrorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.2.9", type: ErrorComponent, isStandalone: true, selector: "hot-toast-error", inputs: { theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "<div\n  class=\"hot-toast-error-icon\"\n  [style.--error-primary]=\"theme()?.primary\"\n  [style.--error-secondary]=\"theme()?.secondary\"\n></div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: ErrorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-error', changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, template: "<div\n  class=\"hot-toast-error-icon\"\n  [style.--error-primary]=\"theme()?.primary\"\n  [style.--error-secondary]=\"theme()?.secondary\"\n></div>\n" }]
        }], propDecorators: { theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }] } });

class CheckMarkComponent {
    constructor() {
        this.theme = input(...(ngDevMode ? [undefined, { debugName: "theme" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: CheckMarkComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.2.9", type: CheckMarkComponent, isStandalone: true, selector: "hot-toast-checkmark", inputs: { theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "<div\n  class=\"hot-toast-checkmark-icon\"\n  [style.--check-primary]=\"theme()?.primary\"\n  [style.--check-secondary]=\"theme()?.secondary\"\n></div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: CheckMarkComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-checkmark', changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, template: "<div\n  class=\"hot-toast-checkmark-icon\"\n  [style.--check-primary]=\"theme()?.primary\"\n  [style.--check-secondary]=\"theme()?.secondary\"\n></div>\n" }]
        }], propDecorators: { theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }] } });

class WarningComponent {
    constructor() {
        this.theme = input(...(ngDevMode ? [undefined, { debugName: "theme" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: WarningComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.2.9", type: WarningComponent, isStandalone: true, selector: "hot-toast-warning", inputs: { theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "<div\n  class=\"hot-toast-warning-icon\"\n  [style.--warn-primary]=\"theme()?.primary\"\n  [style.--warn-secondary]=\"theme()?.secondary\"\n></div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: WarningComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-warning', changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, template: "<div\n  class=\"hot-toast-warning-icon\"\n  [style.--warn-primary]=\"theme()?.primary\"\n  [style.--warn-secondary]=\"theme()?.secondary\"\n></div>\n" }]
        }], propDecorators: { theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }] } });

class InfoComponent {
    constructor() {
        this.theme = input(...(ngDevMode ? [undefined, { debugName: "theme" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: InfoComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.2.9", type: InfoComponent, isStandalone: true, selector: "hot-toast-info", inputs: { theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "<div\n  class=\"hot-toast-info-icon\"\n  [style.--info-primary]=\"theme()?.primary\"\n  [style.--info-secondary]=\"theme()?.secondary\"\n></div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: InfoComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-info', changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, template: "<div\n  class=\"hot-toast-info-icon\"\n  [style.--info-primary]=\"theme()?.primary\"\n  [style.--info-secondary]=\"theme()?.secondary\"\n></div>\n" }]
        }], propDecorators: { theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }] } });

class IndicatorComponent {
    constructor() {
        this.theme = input(...(ngDevMode ? [undefined, { debugName: "theme" }] : /* istanbul ignore next */ []));
        this.type = input(...(ngDevMode ? [undefined, { debugName: "type" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: IndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.2.9", type: IndicatorComponent, isStandalone: true, selector: "hot-toast-indicator", inputs: { theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null }, type: { classPropertyName: "type", publicName: "type", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "@if (type() !== 'blank') {\n<div class=\"hot-toast-indicator-wrapper\">\n  @if (type() === 'loading') {\n  <hot-toast-loader [theme]=\"theme()\" />\n  } @if (type() !== 'loading') {\n  <div class=\"hot-toast-status-wrapper\">\n    <div>\n      @switch (type()) { @case ('error') {\n      <div>\n        <hot-toast-error [theme]=\"theme()\" />\n      </div>\n      } @case ('success') {\n      <div>\n        <hot-toast-checkmark [theme]=\"theme()\" />\n      </div>\n      } @case ('warning') {\n      <div>\n        <hot-toast-warning [theme]=\"theme()\" />\n      </div>\n      } @case ('info') {\n      <div>\n        <hot-toast-info [theme]=\"theme()\" />\n      </div>\n      } }\n    </div>\n  </div>\n  }\n</div>\n}\n", dependencies: [{ kind: "component", type: LoaderComponent, selector: "hot-toast-loader", inputs: ["theme"] }, { kind: "component", type: ErrorComponent, selector: "hot-toast-error", inputs: ["theme"] }, { kind: "component", type: CheckMarkComponent, selector: "hot-toast-checkmark", inputs: ["theme"] }, { kind: "component", type: WarningComponent, selector: "hot-toast-warning", inputs: ["theme"] }, { kind: "component", type: InfoComponent, selector: "hot-toast-info", inputs: ["theme"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: IndicatorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-indicator', changeDetection: ChangeDetectionStrategy.OnPush, imports: [LoaderComponent, ErrorComponent, CheckMarkComponent, WarningComponent, InfoComponent], template: "@if (type() !== 'blank') {\n<div class=\"hot-toast-indicator-wrapper\">\n  @if (type() === 'loading') {\n  <hot-toast-loader [theme]=\"theme()\" />\n  } @if (type() !== 'loading') {\n  <div class=\"hot-toast-status-wrapper\">\n    <div>\n      @switch (type()) { @case ('error') {\n      <div>\n        <hot-toast-error [theme]=\"theme()\" />\n      </div>\n      } @case ('success') {\n      <div>\n        <hot-toast-checkmark [theme]=\"theme()\" />\n      </div>\n      } @case ('warning') {\n      <div>\n        <hot-toast-warning [theme]=\"theme()\" />\n      </div>\n      } @case ('info') {\n      <div>\n        <hot-toast-info [theme]=\"theme()\" />\n      </div>\n      } }\n    </div>\n  </div>\n  }\n</div>\n}\n" }]
        }], propDecorators: { theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }], type: [{ type: i0.Input, args: [{ isSignal: true, alias: "type", required: false }] }] } });

class AnimatedIconComponent {
    constructor() {
        this.iconTheme = input(...(ngDevMode ? [undefined, { debugName: "iconTheme" }] : /* istanbul ignore next */ []));
        this.icon = input(...(ngDevMode ? [undefined, { debugName: "icon" }] : /* istanbul ignore next */ []));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: AnimatedIconComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "21.2.9", type: AnimatedIconComponent, isStandalone: true, selector: "hot-toast-animated-icon", inputs: { iconTheme: { classPropertyName: "iconTheme", publicName: "iconTheme", isSignal: true, isRequired: false, transformFunction: null }, icon: { classPropertyName: "icon", publicName: "icon", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: "<div class=\"hot-toast-animated-icon\" [style.color]=\"iconTheme()?.primary\">\n  <ng-container *dynamicView=\"icon()\" />\n</div>\n", dependencies: [{ kind: "directive", type: DynamicViewDirective, selector: "[dynamicView]", inputs: ["dynamicView", "dynamicViewInjector", "dynamicViewContext", "dynamicViewInputs"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: AnimatedIconComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-animated-icon', changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, imports: [DynamicViewDirective], template: "<div class=\"hot-toast-animated-icon\" [style.color]=\"iconTheme()?.primary\">\n  <ng-container *dynamicView=\"icon()\" />\n</div>\n" }]
        }], propDecorators: { iconTheme: [{ type: i0.Input, args: [{ isSignal: true, alias: "iconTheme", required: false }] }], icon: [{ type: i0.Input, args: [{ isSignal: true, alias: "icon", required: false }] }] } });

class HotToastGroupItemComponent {
    constructor() {
        this.offset = input(0, ...(ngDevMode ? [{ debugName: "offset" }] : /* istanbul ignore next */ []));
        this.defaultConfig = input(...(ngDevMode ? [undefined, { debugName: "defaultConfig" }] : /* istanbul ignore next */ []));
        this.toastRef = input(...(ngDevMode ? [undefined, { debugName: "toastRef" }] : /* istanbul ignore next */ []));
        this._toastsAfter = 0;
        this.isShowingAllToasts = input(false, ...(ngDevMode ? [{ debugName: "isShowingAllToasts" }] : /* istanbul ignore next */ []));
        this.height = output();
        this.beforeClosed = output();
        this.afterClosed = output();
        this.showAllToasts = output();
        this.toggleGroup = output();
        this.isManualClose = false;
        this.toastBarBaseStylesSignal = signal({}, ...(ngDevMode ? [{ debugName: "toastBarBaseStylesSignal" }] : /* istanbul ignore next */ []));
        this.unlisteners = [];
        this.rafIds = [];
        this.destroyed = false;
        this.softClosed = false;
        this.injector = inject(Injector);
        this.renderer = inject(Renderer2);
        this.ngZone = inject(NgZone);
        this.cdr = inject(ChangeDetectorRef);
    }
    set toast(value) {
        this._toast = value;
        const ogStyle = this.toastBarBaseStylesSignal();
        const newStyle = { ...value.style };
        if (ogStyle['animation']?.includes('hotToastExitAnimation')) {
            // if toast is set for exit, we don't need want set the enter animation
            newStyle['animation'] = ogStyle['animation'];
        }
        else {
            const top = value.position.includes('top');
            const enterAnimation = `hotToastEnterAnimation${top ? 'Negative' : 'Positive'} ${ENTER_ANIMATION_DURATION}ms cubic-bezier(0.21, 1.02, 0.73, 1) forwards`;
            newStyle['animation'] = enterAnimation;
        }
        this.toastBarBaseStylesSignal.set(newStyle);
    }
    get toast() {
        return this._toast;
    }
    get toastsAfter() {
        return this._toastsAfter;
    }
    set toastsAfter(value) {
        this._toastsAfter = value;
    }
    get toastBarBaseHeight() {
        return this.toastBarBase.nativeElement.offsetHeight;
    }
    get scale() {
        return this.defaultConfig().stacking !== 'vertical' && !this.isShowingAllToasts()
            ? this.toastsAfter * -HOT_TOAST_DEPTH_SCALE + 1
            : 1;
    }
    get translateY() {
        return this.offset() * (this.top ? 1 : -1) + 'px';
    }
    get exitAnimationDelay() {
        return this.toast.duration + 'ms';
    }
    get top() {
        return this.toast.position.includes('top');
    }
    get containerPositionStyle() {
        const verticalStyle = this.top ? { top: 0 } : { bottom: 0 };
        const transform = `translateY(var(--hot-toast-translate-y)) scale(var(--hot-toast-scale))`;
        const horizontalStyle = this.toast.position.includes('left')
            ? {
                left: 0,
            }
            : this.toast.position.includes('right')
                ? {
                    right: 0,
                }
                : {
                    left: 0,
                    right: 0,
                    justifyContent: 'center',
                };
        return {
            transform,
            ...verticalStyle,
            ...horizontalStyle,
        };
    }
    get isIconString() {
        return typeof this.toast.icon === 'string';
    }
    get groupChildrenToastRefs() {
        return this.toastRef().groupRefs.filter((ref) => !!ref);
    }
    set groupChildrenToastRefs(value) {
        this.toastRef().groupRefs = value;
    }
    get groupChildrenToasts() {
        return this.groupChildrenToastRefs.map((ref) => ref.getToast());
    }
    get groupHeight() {
        return this.visibleToasts.map((t) => t.height).reduce((prev, curr) => prev + curr, 0);
    }
    get isExpanded() {
        return this.toastRef().groupExpanded;
    }
    ngOnChanges(changes) {
        if (changes.toast && !changes.toast.firstChange && changes.toast.currentValue?.message) {
            const rafId = requestAnimationFrame(() => {
                this.rafIds = this.rafIds.filter((id) => id !== rafId);
                if (this.destroyed)
                    return;
                this.height.emit(this.toastBarBase.nativeElement.offsetHeight);
            });
            this.rafIds.push(rafId);
        }
    }
    ngOnInit() {
        if (isTemplateRef(this.toast.message)) {
            this.context = { $implicit: this.toastRef() };
        }
        if (isComponent(this.toast.message)) {
            this.toastComponentInjector = Injector.create({
                providers: [
                    {
                        provide: HotToastRef,
                        useValue: this.toastRef(),
                    },
                ],
                parent: this.toast.injector || this.injector,
            });
        }
        const nativeElement = this.toastBarBase.nativeElement;
        // Caretaker note: `animationstart` and `animationend` events are event tasks that trigger change detection.
        // We'd want to trigger the change detection only if it's an exit animation.
        this.ngZone.runOutsideAngular(() => {
            this.unlisteners.push(
            // Caretaker note: we have to remove these event listeners at the end (even if the element is removed from DOM).
            this.renderer.listen(nativeElement, 'animationstart', (event) => {
                if (this.isExitAnimation(event)) {
                    this.ngZone.run(() => {
                        this.renderer.setStyle(nativeElement, 'pointer-events', 'none');
                        this.renderer.setStyle(nativeElement.parentElement, 'pointer-events', 'none');
                        this.beforeClosed.emit();
                    });
                }
            }), this.renderer.listen(nativeElement, 'animationend', (event) => {
                if (this.isEnterAnimation(event)) {
                    this.ngZone.run(() => {
                        if (this.toast.autoClose) {
                            const exitAnimation = `hotToastExitAnimation${this.top ? 'Negative' : 'Positive'} ${EXIT_ANIMATION_DURATION}ms forwards cubic-bezier(0.06, 0.71, 0.55, 1) var(--hot-toast-exit-animation-delay) var(--hot-toast-exit-animation-state)`;
                            this.toastBarBaseStylesSignal.set({ ...this.toast.style, animation: exitAnimation });
                        }
                    });
                }
                if (this.isExitAnimation(event)) {
                    this.ngZone.run(() => this.afterClosed.emit({ dismissedByAction: this.isManualClose, id: this.toast.id }));
                }
            }));
        });
    }
    ngAfterViewInit() {
        const nativeElement = this.toastBarBase.nativeElement;
        // Caretaker note: accessing `offsetHeight` triggers the whole layout update.
        // Macro tasks (like `setTimeout`) might be executed within the current rendering frame and cause a frame drop.
        const rafId = requestAnimationFrame(() => {
            this.rafIds = this.rafIds.filter((id) => id !== rafId);
            if (this.destroyed)
                return;
            this.height.emit(nativeElement.offsetHeight);
        });
        this.rafIds.push(rafId);
        this.setToastAttributes();
    }
    softClose() {
        const exitAnimation = `hotToastExitSoftAnimation${this.top ? 'Negative' : 'Positive'} ${EXIT_ANIMATION_DURATION}ms forwards cubic-bezier(0.06, 0.71, 0.55, 1)`;
        const nativeElement = this.toastBarBase.nativeElement;
        animate(this.renderer, nativeElement, exitAnimation);
        this.softClosed = true;
    }
    softOpen() {
        const softEnterAnimation = `hotToastEnterSoftAnimation${top ? 'Negative' : 'Positive'} ${ENTER_ANIMATION_DURATION}ms cubic-bezier(0.21, 1.02, 0.73, 1) forwards`;
        const nativeElement = this.toastBarBase.nativeElement;
        animate(this.renderer, nativeElement, softEnterAnimation);
        this.softClosed = false;
    }
    close() {
        this.isManualClose = true;
        this.cdr.markForCheck();
        const exitAnimation = `hotToastExitAnimation${this.top ? 'Negative' : 'Positive'} ${EXIT_ANIMATION_DURATION}ms forwards cubic-bezier(0.06, 0.71, 0.55, 1)`;
        this.toastBarBaseStylesSignal.set({ ...this.toast.style, animation: exitAnimation });
    }
    handleMouseEnter() {
        this.showAllToasts.emit(true);
    }
    handleMouseLeave() {
        this.showAllToasts.emit(false);
    }
    ngOnDestroy() {
        this.destroyed = true;
        while (this.rafIds.length) {
            cancelAnimationFrame(this.rafIds.pop());
        }
        this.close();
        while (this.unlisteners.length) {
            this.unlisteners.pop()();
        }
    }
    isExitAnimation(ev) {
        return ev.animationName.includes('hotToastExitAnimation');
    }
    isEnterAnimation(ev) {
        return ev.animationName.includes('hotToastEnterAnimation');
    }
    setToastAttributes() {
        const toastAttributes = this.toast.attributes;
        for (const [key, value] of Object.entries(toastAttributes)) {
            this.renderer.setAttribute(this.toastBarBase.nativeElement, key, value);
        }
    }
    get visibleToasts() {
        return this.groupChildrenToasts.filter((t) => t.visible);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastGroupItemComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.2.9", type: HotToastGroupItemComponent, isStandalone: true, selector: "hot-toast-group-item", inputs: { toast: { classPropertyName: "toast", publicName: "toast", isSignal: false, isRequired: false, transformFunction: null }, offset: { classPropertyName: "offset", publicName: "offset", isSignal: true, isRequired: false, transformFunction: null }, defaultConfig: { classPropertyName: "defaultConfig", publicName: "defaultConfig", isSignal: true, isRequired: false, transformFunction: null }, toastRef: { classPropertyName: "toastRef", publicName: "toastRef", isSignal: true, isRequired: false, transformFunction: null }, toastsAfter: { classPropertyName: "toastsAfter", publicName: "toastsAfter", isSignal: false, isRequired: false, transformFunction: null }, isShowingAllToasts: { classPropertyName: "isShowingAllToasts", publicName: "isShowingAllToasts", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { height: "height", beforeClosed: "beforeClosed", afterClosed: "afterClosed", showAllToasts: "showAllToasts", toggleGroup: "toggleGroup" }, viewQueries: [{ propertyName: "toastBarBase", first: true, predicate: ["hotToastBarBase"], descendants: true, static: true }], usesOnChanges: true, ngImport: i0, template: "<div\n  class=\"hot-toast-bar-base-container\"\n  [style]=\"containerPositionStyle\"\n  [class]=\"'hot-toast-theme-' + toast.theme\"\n  [style.--hot-toast-scale]=\"scale\"\n  [style.--hot-toast-translate-y]=\"translateY\"\n>\n  <div class=\"hot-toast-bar-base-wrapper\" (mouseenter)=\"handleMouseEnter()\" (mouseleave)=\"handleMouseLeave()\">\n    <div\n      class=\"hot-toast-bar-base\"\n      #hotToastBarBase\n      [style]=\"toastBarBaseStylesSignal()\"\n      [class]=\"toast.className\"\n      [style.--hot-toast-animation-state]=\"isManualClose ? 'running' : 'paused'\"\n      [style.--hot-toast-exit-animation-state]=\"isShowingAllToasts() ? 'paused' : 'running'\"\n      [style.--hot-toast-exit-animation-delay]=\"exitAnimationDelay\"\n      [attr.aria-live]=\"toast.ariaLive\"\n      [attr.role]=\"toast.role\"\n    >\n      <div class=\"hot-toast-icon\" aria-hidden=\"true\">\n        @if (toast.icon !== undefined) { @if (isIconString) {\n        <hot-toast-animated-icon [iconTheme]=\"toast.iconTheme\">{{ toast.icon }}</hot-toast-animated-icon>\n        } @else {\n        <div>\n          <ng-container *dynamicView=\"toast.icon\" />\n        </div>\n        } } @else {\n        <hot-toast-indicator [theme]=\"toast.iconTheme\" [type]=\"toast.type\" />\n        }\n      </div>\n      <div class=\"hot-toast-message\">\n        <ng-container *dynamicView=\"toast.message; context: context; injector: toastComponentInjector\" />\n      </div>\n      @if (toast.dismissible) {\n      <button\n        (click)=\"close()\"\n        type=\"button\"\n        class=\"hot-toast-close-btn\"\n        aria-label=\"Close\"\n        [style]=\"toast.closeStyle\"\n      ></button>\n      }\n    </div>\n  </div>\n</div>\n", dependencies: [{ kind: "component", type: AnimatedIconComponent, selector: "hot-toast-animated-icon", inputs: ["iconTheme", "icon"] }, { kind: "component", type: IndicatorComponent, selector: "hot-toast-indicator", inputs: ["theme", "type"] }, { kind: "directive", type: DynamicViewDirective, selector: "[dynamicView]", inputs: ["dynamicView", "dynamicViewInjector", "dynamicViewContext", "dynamicViewInputs"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastGroupItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-group-item', changeDetection: ChangeDetectionStrategy.OnPush, imports: [AnimatedIconComponent, IndicatorComponent, DynamicViewDirective], template: "<div\n  class=\"hot-toast-bar-base-container\"\n  [style]=\"containerPositionStyle\"\n  [class]=\"'hot-toast-theme-' + toast.theme\"\n  [style.--hot-toast-scale]=\"scale\"\n  [style.--hot-toast-translate-y]=\"translateY\"\n>\n  <div class=\"hot-toast-bar-base-wrapper\" (mouseenter)=\"handleMouseEnter()\" (mouseleave)=\"handleMouseLeave()\">\n    <div\n      class=\"hot-toast-bar-base\"\n      #hotToastBarBase\n      [style]=\"toastBarBaseStylesSignal()\"\n      [class]=\"toast.className\"\n      [style.--hot-toast-animation-state]=\"isManualClose ? 'running' : 'paused'\"\n      [style.--hot-toast-exit-animation-state]=\"isShowingAllToasts() ? 'paused' : 'running'\"\n      [style.--hot-toast-exit-animation-delay]=\"exitAnimationDelay\"\n      [attr.aria-live]=\"toast.ariaLive\"\n      [attr.role]=\"toast.role\"\n    >\n      <div class=\"hot-toast-icon\" aria-hidden=\"true\">\n        @if (toast.icon !== undefined) { @if (isIconString) {\n        <hot-toast-animated-icon [iconTheme]=\"toast.iconTheme\">{{ toast.icon }}</hot-toast-animated-icon>\n        } @else {\n        <div>\n          <ng-container *dynamicView=\"toast.icon\" />\n        </div>\n        } } @else {\n        <hot-toast-indicator [theme]=\"toast.iconTheme\" [type]=\"toast.type\" />\n        }\n      </div>\n      <div class=\"hot-toast-message\">\n        <ng-container *dynamicView=\"toast.message; context: context; injector: toastComponentInjector\" />\n      </div>\n      @if (toast.dismissible) {\n      <button\n        (click)=\"close()\"\n        type=\"button\"\n        class=\"hot-toast-close-btn\"\n        aria-label=\"Close\"\n        [style]=\"toast.closeStyle\"\n      ></button>\n      }\n    </div>\n  </div>\n</div>\n" }]
        }], propDecorators: { toast: [{
                type: Input
            }], offset: [{ type: i0.Input, args: [{ isSignal: true, alias: "offset", required: false }] }], defaultConfig: [{ type: i0.Input, args: [{ isSignal: true, alias: "defaultConfig", required: false }] }], toastRef: [{ type: i0.Input, args: [{ isSignal: true, alias: "toastRef", required: false }] }], toastsAfter: [{
                type: Input
            }], isShowingAllToasts: [{ type: i0.Input, args: [{ isSignal: true, alias: "isShowingAllToasts", required: false }] }], height: [{ type: i0.Output, args: ["height"] }], beforeClosed: [{ type: i0.Output, args: ["beforeClosed"] }], afterClosed: [{ type: i0.Output, args: ["afterClosed"] }], showAllToasts: [{ type: i0.Output, args: ["showAllToasts"] }], toggleGroup: [{ type: i0.Output, args: ["toggleGroup"] }], toastBarBase: [{
                type: ViewChild,
                args: ['hotToastBarBase', { static: true }]
            }] } });

class HotToastComponent {
    constructor() {
        this.offset = input(0, ...(ngDevMode ? [{ debugName: "offset" }] : /* istanbul ignore next */ []));
        this.defaultConfig = input(...(ngDevMode ? [undefined, { debugName: "defaultConfig" }] : /* istanbul ignore next */ []));
        this.toastRef = input(...(ngDevMode ? [undefined, { debugName: "toastRef" }] : /* istanbul ignore next */ []));
        this._toastsAfter = 0;
        this.isShowingAllToasts = input(false, ...(ngDevMode ? [{ debugName: "isShowingAllToasts" }] : /* istanbul ignore next */ []));
        this.height = output();
        this.beforeClosed = output();
        this.afterClosed = output();
        this.showAllToasts = output();
        this.toggleGroup = output();
        this.isManualClose = false;
        this.isExpanded = false;
        this.toastBarBaseStylesSignal = signal({}, ...(ngDevMode ? [{ debugName: "toastBarBaseStylesSignal" }] : /* istanbul ignore next */ []));
        this.unlisteners = [];
        this.rafIds = [];
        this.destroyed = false;
        this.softClosed = false;
        this.groupRefs = [];
        this.injector = inject(Injector);
        this.renderer = inject(Renderer2);
        this.ngZone = inject(NgZone);
        this.cdr = inject(ChangeDetectorRef);
    }
    set toast(value) {
        this._toast = value;
        const ogStyle = this.toastBarBaseStylesSignal();
        const newStyle = { ...value.style };
        if (ogStyle['animation']?.includes('hotToastExitAnimation')) {
            // if toast is set for exit, we don't need want set the enter animation
            newStyle['animation'] = ogStyle['animation'];
        }
        else {
            const top = value.position.includes('top');
            const enterAnimation = `hotToastEnterAnimation${top ? 'Negative' : 'Positive'} ${ENTER_ANIMATION_DURATION}ms cubic-bezier(0.21, 1.02, 0.73, 1) forwards`;
            newStyle['animation'] = enterAnimation;
        }
        this.toastBarBaseStylesSignal.set(newStyle);
    }
    get toast() {
        return this._toast;
    }
    get toastsAfter() {
        return this._toastsAfter;
    }
    set toastsAfter(value) {
        this._toastsAfter = value;
        const defaultConfig = this.defaultConfig();
        if (defaultConfig?.visibleToasts > 0) {
            if (this.toast.autoClose) {
                // if (value >= this.defaultConfig?.visibleToasts) {
                //   this.close();
                // }
            }
            else {
                if (value >= defaultConfig?.visibleToasts) {
                    this.softClose();
                }
                else if (this.softClosed) {
                    this.softOpen();
                }
            }
        }
    }
    get toastBarBaseHeight() {
        return this.toastBarBase.nativeElement.offsetHeight;
    }
    get scale() {
        return this.defaultConfig().stacking !== 'vertical' && !this.isShowingAllToasts()
            ? this.toastsAfter * -HOT_TOAST_DEPTH_SCALE + 1
            : 1;
    }
    get translateY() {
        return this.offset() * (this.top ? 1 : -1) + 'px';
    }
    get exitAnimationDelay() {
        return this.toast.duration + 'ms';
    }
    get top() {
        return this.toast.position.includes('top');
    }
    get containerPositionStyle() {
        const verticalStyle = this.top ? { top: 0 } : { bottom: 0 };
        const transform = `translateY(var(--hot-toast-translate-y)) scale(var(--hot-toast-scale))`;
        const horizontalStyle = this.toast.position.includes('left')
            ? {
                left: 0,
            }
            : this.toast.position.includes('right')
                ? {
                    right: 0,
                }
                : {
                    left: 0,
                    right: 0,
                    justifyContent: 'center',
                };
        return {
            transform,
            ...verticalStyle,
            ...horizontalStyle,
        };
    }
    get isIconString() {
        return typeof this.toast.icon === 'string';
    }
    get groupChildrenToastRefs() {
        return this.groupRefs.filter((ref) => !!ref);
    }
    set groupChildrenToastRefs(value) {
        this.groupRefs = value;
        this.toastRef().groupRefs = value;
    }
    get groupChildrenToasts() {
        return this.groupChildrenToastRefs.map((ref) => ref.getToast());
    }
    get groupHeight() {
        return this.visibleToasts
            .slice(-this.defaultConfig().visibleToasts)
            .map((t) => t.height)
            .reduce((prev, curr) => prev + curr, 0);
    }
    get visibleToasts() {
        return this.groupChildrenToasts.filter((t) => t.visible);
    }
    ngDoCheck() {
        const toastRef = this.toastRef();
        if (toastRef.groupRefs.length !== this.groupRefs.length) {
            this.groupRefs = toastRef.groupRefs.slice();
            this.cdr.markForCheck();
            this.emiHeightWithGroup(this.isExpanded);
        }
        if (toastRef.groupExpanded !== this.isExpanded) {
            this.isExpanded = toastRef.groupExpanded;
            this.cdr.markForCheck();
            this.emiHeightWithGroup(this.isExpanded);
        }
    }
    ngOnChanges(changes) {
        if (changes.toast && !changes.toast.firstChange && changes.toast.currentValue?.message) {
            (this, this.emiHeightWithGroup(this.isExpanded));
        }
    }
    ngOnInit() {
        if (isTemplateRef(this.toast.message)) {
            this.context = { $implicit: this.toastRef() };
        }
        if (isComponent(this.toast.message)) {
            this.toastComponentInjector = Injector.create({
                providers: [
                    {
                        provide: HotToastRef,
                        useValue: this.toastRef(),
                    },
                ],
                parent: this.toast.injector || this.injector,
            });
        }
        const nativeElement = this.toastBarBase.nativeElement;
        // Caretaker note: `animationstart` and `animationend` events are event tasks that trigger change detection.
        // We'd want to trigger the change detection only if it's an exit animation.
        this.ngZone.runOutsideAngular(() => {
            this.unlisteners.push(
            // Caretaker note: we have to remove these event listeners at the end (even if the element is removed from DOM).
            this.renderer.listen(nativeElement, 'animationstart', (event) => {
                if (this.isExitAnimation(event)) {
                    this.ngZone.run(() => {
                        this.renderer.setStyle(nativeElement, 'pointer-events', 'none');
                        this.renderer.setStyle(nativeElement.parentElement, 'pointer-events', 'none');
                        this.beforeClosed.emit();
                    });
                }
            }), this.renderer.listen(nativeElement, 'animationend', (event) => {
                if (this.isEnterAnimation(event)) {
                    this.ngZone.run(() => {
                        if (this.toast.autoClose) {
                            const exitAnimation = `hotToastExitAnimation${this.top ? 'Negative' : 'Positive'} ${EXIT_ANIMATION_DURATION}ms forwards cubic-bezier(0.06, 0.71, 0.55, 1) var(--hot-toast-exit-animation-delay) var(--hot-toast-exit-animation-state)`;
                            this.toastBarBaseStylesSignal.set({ ...this.toast.style, animation: exitAnimation });
                        }
                    });
                }
                if (this.isExitAnimation(event)) {
                    this.ngZone.run(() => this.afterClosed.emit({ dismissedByAction: this.isManualClose, id: this.toast.id }));
                }
            }));
        });
    }
    ngAfterViewInit() {
        const nativeElement = this.toastBarBase.nativeElement;
        // Caretaker note: accessing `offsetHeight` triggers the whole layout update.
        // Macro tasks (like `setTimeout`) might be executed within the current rendering frame and cause a frame drop.
        const rafId = requestAnimationFrame(() => {
            this.rafIds = this.rafIds.filter((id) => id !== rafId);
            if (this.destroyed)
                return;
            this.height.emit(nativeElement.offsetHeight);
        });
        this.rafIds.push(rafId);
        this.setToastAttributes();
    }
    softClose() {
        const exitAnimation = `hotToastExitSoftAnimation${this.top ? 'Negative' : 'Positive'} ${EXIT_ANIMATION_DURATION}ms forwards cubic-bezier(0.06, 0.71, 0.55, 1)`;
        const nativeElement = this.toastBarBase.nativeElement;
        animate(this.renderer, nativeElement, exitAnimation);
        this.softClosed = true;
        if (this.isExpanded) {
            this.toggleToastGroup();
        }
    }
    softOpen() {
        const softEnterAnimation = `hotToastEnterSoftAnimation${top ? 'Negative' : 'Positive'} ${ENTER_ANIMATION_DURATION}ms cubic-bezier(0.21, 1.02, 0.73, 1) forwards`;
        const nativeElement = this.toastBarBase.nativeElement;
        animate(this.renderer, nativeElement, softEnterAnimation);
        this.softClosed = false;
    }
    close() {
        this.isManualClose = true;
        this.cdr.markForCheck();
        const exitAnimation = `hotToastExitAnimation${this.top ? 'Negative' : 'Positive'} ${EXIT_ANIMATION_DURATION}ms forwards cubic-bezier(0.06, 0.71, 0.55, 1)`;
        this.toastBarBaseStylesSignal.set({ ...this.toast.style, animation: exitAnimation });
    }
    handleMouseEnter() {
        this.showAllToasts.emit(true);
    }
    handleMouseLeave() {
        this.showAllToasts.emit(false);
    }
    ngOnDestroy() {
        this.destroyed = true;
        while (this.rafIds.length) {
            cancelAnimationFrame(this.rafIds.pop());
        }
        this.close();
        while (this.unlisteners.length) {
            this.unlisteners.pop()();
        }
    }
    isExitAnimation(ev) {
        return ev.animationName.includes('hotToastExitAnimation');
    }
    isEnterAnimation(ev) {
        return ev.animationName.includes('hotToastEnterAnimation');
    }
    setToastAttributes() {
        const toastAttributes = this.toast.attributes;
        for (const [key, value] of Object.entries(toastAttributes)) {
            this.renderer.setAttribute(this.toastBarBase.nativeElement, key, value);
        }
    }
    calculateOffset(toastId) {
        const visibleToasts = this.visibleToasts;
        const index = visibleToasts.findIndex((toast) => toast.id === toastId);
        const offset = index !== -1
            ? visibleToasts.slice(...(this.defaultConfig().reverseOrder ? [index + 1] : [0, index])).reduce((acc, t, i) => {
                const defaultConfig = this.defaultConfig();
                return defaultConfig.visibleToasts !== 0 && i < visibleToasts.length - defaultConfig.visibleToasts
                    ? 0
                    : acc + (t.height || 0);
            }, 0)
            : 0;
        return offset;
    }
    updateHeight(height, toast) {
        toast.height = height;
        this.cdr.markForCheck();
    }
    beforeClosedGroupItem(toast) {
        toast.visible = false;
        this.cdr.markForCheck();
        if (this.visibleToasts.length === 0 && this.isExpanded) {
            this.toggleToastGroup();
        }
        else {
            this.emiHeightWithGroup(this.isExpanded);
        }
    }
    afterClosedGroupItem(closeToast) {
        const toastIndex = this.groupChildrenToasts.findIndex((t) => t.id === closeToast.id);
        if (toastIndex > -1) {
            this.groupChildrenToastRefs = this.groupChildrenToastRefs.filter((t) => t.getToast().id !== closeToast.id);
            this.cdr.markForCheck();
        }
    }
    toggleToastGroup() {
        const event = this.isExpanded ? 'collapse' : 'expand';
        this.toggleGroup.emit({
            byAction: true,
            event,
            id: this.toast.id,
        });
        this.emiHeightWithGroup(event === 'expand');
    }
    emiHeightWithGroup(isExpanded) {
        if (isExpanded) {
            const rafId = requestAnimationFrame(() => {
                this.rafIds = this.rafIds.filter((id) => id !== rafId);
                if (this.destroyed)
                    return;
                this.height.emit(this.toastBarBase.nativeElement.offsetHeight + this.groupHeight);
            });
            this.rafIds.push(rafId);
        }
        else {
            const rafId = requestAnimationFrame(() => {
                this.rafIds = this.rafIds.filter((id) => id !== rafId);
                if (this.destroyed)
                    return;
                this.height.emit(this.toastBarBase.nativeElement.offsetHeight);
            });
            this.rafIds.push(rafId);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.2.9", type: HotToastComponent, isStandalone: true, selector: "hot-toast-component", inputs: { toast: { classPropertyName: "toast", publicName: "toast", isSignal: false, isRequired: false, transformFunction: null }, offset: { classPropertyName: "offset", publicName: "offset", isSignal: true, isRequired: false, transformFunction: null }, defaultConfig: { classPropertyName: "defaultConfig", publicName: "defaultConfig", isSignal: true, isRequired: false, transformFunction: null }, toastRef: { classPropertyName: "toastRef", publicName: "toastRef", isSignal: true, isRequired: false, transformFunction: null }, toastsAfter: { classPropertyName: "toastsAfter", publicName: "toastsAfter", isSignal: false, isRequired: false, transformFunction: null }, isShowingAllToasts: { classPropertyName: "isShowingAllToasts", publicName: "isShowingAllToasts", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { height: "height", beforeClosed: "beforeClosed", afterClosed: "afterClosed", showAllToasts: "showAllToasts", toggleGroup: "toggleGroup" }, viewQueries: [{ propertyName: "toastBarBase", first: true, predicate: ["hotToastBarBase"], descendants: true, static: true }], usesOnChanges: true, ngImport: i0, template: "<div\n  class=\"hot-toast-bar-base-container\"\n  [style]=\"containerPositionStyle\"\n  [class]=\"'hot-toast-theme-' + toast.theme\"\n  [style.--hot-toast-scale]=\"scale\"\n  [style.--hot-toast-translate-y]=\"translateY\"\n>\n  <div\n    class=\"hot-toast-bar-base-wrapper\"\n    [class.expanded]=\"isExpanded\"\n    (mouseenter)=\"handleMouseEnter()\"\n    (mouseleave)=\"handleMouseLeave()\"\n  >\n    <div\n      class=\"hot-toast-bar-base\"\n      #hotToastBarBase\n      [style]=\"toastBarBaseStylesSignal()\"\n      [class]=\"toast.className\"\n      [style.--hot-toast-animation-state]=\"isManualClose ? 'running' : 'paused'\"\n      [style.--hot-toast-exit-animation-state]=\"isShowingAllToasts() ? 'paused' : 'running'\"\n      [style.--hot-toast-exit-animation-delay]=\"exitAnimationDelay\"\n      [attr.aria-live]=\"toast.ariaLive\"\n      [attr.role]=\"toast.role\"\n    >\n      <div class=\"hot-toast-icon\" aria-hidden=\"true\">\n        @if (toast.icon !== undefined) { @if (isIconString) {\n        <hot-toast-animated-icon [iconTheme]=\"toast.iconTheme\" [icon]=\"toast.icon\" />\n        } @else {\n        <div>\n          <ng-container *dynamicView=\"toast.icon\" />\n        </div>\n        } } @else {\n        <hot-toast-indicator [theme]=\"toast.iconTheme\" [type]=\"toast.type\" />\n        }\n      </div>\n\n      <div class=\"hot-toast-message\">\n        <ng-container *dynamicView=\"toast.message; context: context; injector: toastComponentInjector\" />\n      </div>\n\n      @if (toast.group?.expandAndCollapsible && toast.group?.children && visibleToasts.length > 0) {\n      <button\n        (click)=\"toggleToastGroup()\"\n        type=\"button\"\n        class=\"hot-toast-group-btn\"\n        [class.expanded]=\"isExpanded\"\n        [attr.aria-label]=\"isExpanded ? 'Collapse' : 'Expand'\"\n        [style]=\"toast.group.btnStyle\"\n      ></button>\n      } @if (toast.dismissible) {\n      <button\n        (click)=\"close()\"\n        type=\"button\"\n        class=\"hot-toast-close-btn\"\n        aria-label=\"Close\"\n        [style]=\"toast.closeStyle\"\n      ></button>\n      }\n    </div>\n\n    @if (toast.visible) {\n    <div\n      role=\"list\"\n      class=\"hot-toast-bar-base-group\"\n      [class]=\"toast.group?.className\"\n      [style.--hot-toast-group-height]=\"groupHeight + 'px'\"\n    >\n      @for (item of groupChildrenToasts; track item.id) {\n      <hot-toast-group-item [toast]=\"item\"\n        [offset]=\"calculateOffset(item.id)\"\n        [toastRef]=\"toastRef().groupRefs[$index]\"\n        [toastsAfter]=\"(item.autoClose ? groupChildrenToasts.length : visibleToasts.length) - 1 - $index\"\n        [defaultConfig]=\"defaultConfig()\"\n        [isShowingAllToasts]=\"isShowingAllToasts()\"\n        (height)=\"updateHeight($event, item)\"\n        (beforeClosed)=\"beforeClosedGroupItem(item)\"\n        (afterClosed)=\"afterClosedGroupItem($event)\"\n       />\n      }\n    </div>\n    }\n  </div>\n</div>\n", dependencies: [{ kind: "directive", type: DynamicViewDirective, selector: "[dynamicView]", inputs: ["dynamicView", "dynamicViewInjector", "dynamicViewContext", "dynamicViewInputs"] }, { kind: "component", type: IndicatorComponent, selector: "hot-toast-indicator", inputs: ["theme", "type"] }, { kind: "component", type: AnimatedIconComponent, selector: "hot-toast-animated-icon", inputs: ["iconTheme", "icon"] }, { kind: "component", type: HotToastGroupItemComponent, selector: "hot-toast-group-item", inputs: ["toast", "offset", "defaultConfig", "toastRef", "toastsAfter", "isShowingAllToasts"], outputs: ["height", "beforeClosed", "afterClosed", "showAllToasts", "toggleGroup"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-component', changeDetection: ChangeDetectionStrategy.OnPush, imports: [DynamicViewDirective, IndicatorComponent, AnimatedIconComponent, HotToastGroupItemComponent], template: "<div\n  class=\"hot-toast-bar-base-container\"\n  [style]=\"containerPositionStyle\"\n  [class]=\"'hot-toast-theme-' + toast.theme\"\n  [style.--hot-toast-scale]=\"scale\"\n  [style.--hot-toast-translate-y]=\"translateY\"\n>\n  <div\n    class=\"hot-toast-bar-base-wrapper\"\n    [class.expanded]=\"isExpanded\"\n    (mouseenter)=\"handleMouseEnter()\"\n    (mouseleave)=\"handleMouseLeave()\"\n  >\n    <div\n      class=\"hot-toast-bar-base\"\n      #hotToastBarBase\n      [style]=\"toastBarBaseStylesSignal()\"\n      [class]=\"toast.className\"\n      [style.--hot-toast-animation-state]=\"isManualClose ? 'running' : 'paused'\"\n      [style.--hot-toast-exit-animation-state]=\"isShowingAllToasts() ? 'paused' : 'running'\"\n      [style.--hot-toast-exit-animation-delay]=\"exitAnimationDelay\"\n      [attr.aria-live]=\"toast.ariaLive\"\n      [attr.role]=\"toast.role\"\n    >\n      <div class=\"hot-toast-icon\" aria-hidden=\"true\">\n        @if (toast.icon !== undefined) { @if (isIconString) {\n        <hot-toast-animated-icon [iconTheme]=\"toast.iconTheme\" [icon]=\"toast.icon\" />\n        } @else {\n        <div>\n          <ng-container *dynamicView=\"toast.icon\" />\n        </div>\n        } } @else {\n        <hot-toast-indicator [theme]=\"toast.iconTheme\" [type]=\"toast.type\" />\n        }\n      </div>\n\n      <div class=\"hot-toast-message\">\n        <ng-container *dynamicView=\"toast.message; context: context; injector: toastComponentInjector\" />\n      </div>\n\n      @if (toast.group?.expandAndCollapsible && toast.group?.children && visibleToasts.length > 0) {\n      <button\n        (click)=\"toggleToastGroup()\"\n        type=\"button\"\n        class=\"hot-toast-group-btn\"\n        [class.expanded]=\"isExpanded\"\n        [attr.aria-label]=\"isExpanded ? 'Collapse' : 'Expand'\"\n        [style]=\"toast.group.btnStyle\"\n      ></button>\n      } @if (toast.dismissible) {\n      <button\n        (click)=\"close()\"\n        type=\"button\"\n        class=\"hot-toast-close-btn\"\n        aria-label=\"Close\"\n        [style]=\"toast.closeStyle\"\n      ></button>\n      }\n    </div>\n\n    @if (toast.visible) {\n    <div\n      role=\"list\"\n      class=\"hot-toast-bar-base-group\"\n      [class]=\"toast.group?.className\"\n      [style.--hot-toast-group-height]=\"groupHeight + 'px'\"\n    >\n      @for (item of groupChildrenToasts; track item.id) {\n      <hot-toast-group-item [toast]=\"item\"\n        [offset]=\"calculateOffset(item.id)\"\n        [toastRef]=\"toastRef().groupRefs[$index]\"\n        [toastsAfter]=\"(item.autoClose ? groupChildrenToasts.length : visibleToasts.length) - 1 - $index\"\n        [defaultConfig]=\"defaultConfig()\"\n        [isShowingAllToasts]=\"isShowingAllToasts()\"\n        (height)=\"updateHeight($event, item)\"\n        (beforeClosed)=\"beforeClosedGroupItem(item)\"\n        (afterClosed)=\"afterClosedGroupItem($event)\"\n       />\n      }\n    </div>\n    }\n  </div>\n</div>\n" }]
        }], propDecorators: { toast: [{
                type: Input
            }], offset: [{ type: i0.Input, args: [{ isSignal: true, alias: "offset", required: false }] }], defaultConfig: [{ type: i0.Input, args: [{ isSignal: true, alias: "defaultConfig", required: false }] }], toastRef: [{ type: i0.Input, args: [{ isSignal: true, alias: "toastRef", required: false }] }], toastsAfter: [{
                type: Input
            }], isShowingAllToasts: [{ type: i0.Input, args: [{ isSignal: true, alias: "isShowingAllToasts", required: false }] }], height: [{ type: i0.Output, args: ["height"] }], beforeClosed: [{ type: i0.Output, args: ["beforeClosed"] }], afterClosed: [{ type: i0.Output, args: ["afterClosed"] }], showAllToasts: [{ type: i0.Output, args: ["showAllToasts"] }], toggleGroup: [{ type: i0.Output, args: ["toggleGroup"] }], toastBarBase: [{
                type: ViewChild,
                args: ['hotToastBarBase', { static: true }]
            }] } });

class HotToastContainerComponent {
    constructor() {
        this.defaultConfig = input(...(ngDevMode ? [undefined, { debugName: "defaultConfig" }] : /* istanbul ignore next */ []));
        this.toasts = [];
        this.toastRefs = [];
        this.isShowingAllToasts = false;
        /** Subject for notifying the user that the toast has been closed. */
        this._onClosed = new Subject();
        /** Subject for notifying the user that the toast has been expanded or collapsed. */
        this._onGroupToggle = new Subject();
        /** Subject for notifying the user that the group refs have been attached to toast. */
        this._onGroupRefAttached = new Subject();
        this.onClosed$ = this._onClosed.asObservable();
        this.onGroupToggle$ = this._onGroupToggle.asObservable();
        this.onGroupRefAttached$ = this._onGroupRefAttached.asObservable();
        this.cdr = inject(ChangeDetectorRef);
        this.toastService = inject(HotToastService);
        this.host = inject(ElementRef);
        afterNextRender(() => {
            if (this.defaultConfig().usePopover) {
                // We need the try/catch because the browser will throw if the
                // host or any of the parents are outside the DOM. Also note
                // the string access which is there for compatibility with Closure.
                try {
                    this.host.nativeElement['showPopover']();
                }
                catch (error) {
                    if (isDevMode()) {
                        console.error('Error showing popover');
                        console.error(error);
                    }
                }
            }
        });
    }
    trackById(index, toast) {
        return toast.id;
    }
    getVisibleToasts(position) {
        return this.unGroupedToasts.filter((t) => t.visible && t.position === position);
    }
    get unGroupedToasts() {
        return this.toasts.filter((t) => t.group?.parent === undefined || t.group?.children === undefined || t.group?.children.length === 0);
    }
    calculateOffset(toastId, position) {
        const visibleToasts = this.getVisibleToasts(position);
        const index = visibleToasts.findIndex((toast) => toast.id === toastId);
        const offset = index !== -1
            ? visibleToasts.slice(...(this.defaultConfig().reverseOrder ? [index + 1] : [0, index])).reduce((acc, t, i) => {
                const toastsAfter = visibleToasts.length - 1 - i;
                return this.defaultConfig().visibleToasts !== 0 &&
                    i < visibleToasts.length - this.defaultConfig().visibleToasts
                    ? 0
                    : acc +
                        (this.defaultConfig().stacking === 'vertical' || this.isShowingAllToasts
                            ? t.height || 0
                            : toastsAfter * HOT_TOAST_DEPTH_SCALE + HOT_TOAST_DEPTH_SCALE_ADD) +
                        HOT_TOAST_MARGIN;
            }, 0)
            : 0;
        return offset;
    }
    updateHeight(height, toast) {
        toast.height = height;
        this.cdr.markForCheck();
    }
    addToast(ref, skipAttachToParent) {
        this.toastRefs.push(ref);
        let toast = ref.getToast();
        this.toasts.push(ref.getToast());
        if (this.defaultConfig().visibleToasts !== 0 && this.unGroupedToasts.length > this.defaultConfig().visibleToasts) {
            const closeToasts = this.toasts.slice(0, this.toasts.length - this.defaultConfig().visibleToasts);
            closeToasts.forEach((t) => {
                if (t.autoClose) {
                    this.closeToast(t.id);
                }
            });
        }
        this.cdr.markForCheck();
        this.attachGroupRefs(toast, ref, skipAttachToParent);
        return {
            dispose: () => {
                this.closeToast(toast.id);
            },
            updateMessage: (message) => {
                toast.message = message;
                this.updateToasts(toast);
                this.cdr.markForCheck();
            },
            updateToast: (options) => {
                toast = { ...toast, ...options };
                this.updateToasts(toast, options);
                this.cdr.markForCheck();
            },
            afterClosed: this.getAfterClosed(toast),
            afterGroupToggled: this.getAfterGroupToggled(toast),
            afterGroupRefsAttached: this.getAfterGroupRefsAttached(toast).pipe(map((v) => v.groupRefs)),
        };
    }
    async attachGroupRefs(toast, ref, skipAttachToParent) {
        let groupRefs = [];
        if (toast.group) {
            if (toast.group.children) {
                groupRefs = await this.createGroupRefs(toast, ref);
                const toastIndex = this.toastRefs.findIndex((t) => t.getToast().id === toast.id);
                if (toastIndex > -1) {
                    this.toastRefs[toastIndex].groupRefs = groupRefs;
                    this.cdr.markForCheck();
                    this._onGroupRefAttached.next({ groupRefs, id: toast.id });
                }
            }
            else if (toast.group.parent && !skipAttachToParent) {
                const parentToastRef = toast.group.parent;
                const parentToast = parentToastRef.getToast();
                const parentToastRefIndex = this.toastRefs.findIndex((t) => t.getToast().id === parentToast.id);
                const parentToastIndex = this.toasts.findIndex((t) => t.id === parentToast.id);
                if (parentToastRefIndex > -1 && parentToastIndex > -1) {
                    this.toastRefs[parentToastRefIndex].groupRefs.push(ref);
                    const existingGroup = this.toasts[parentToastRefIndex].group ?? {};
                    const existingChildren = this.toasts[parentToastRefIndex].group?.children ?? [];
                    existingChildren.push({ options: { ...toast, type: toast.type, message: toast.message } });
                    existingGroup.children = existingChildren;
                    this.toasts[parentToastRefIndex].group = { ...existingGroup };
                    this.cdr.markForCheck();
                    this._onGroupRefAttached.next({ groupRefs, id: parentToast.id });
                }
            }
        }
    }
    createGroupRefs(toast, ref) {
        const skipAttachToParent = true;
        return new Promise((resolve) => {
            const items = toast.group.children;
            const allPromises = items.map((item) => {
                return new Promise((innerResolve) => {
                    item.options.group = { parent: ref };
                    // We need to give a tick's delay so that IDs are generated properly
                    setTimeout(() => {
                        try {
                            const itemRef = this.toastService.show(item.options.message, item.options, skipAttachToParent);
                            innerResolve(itemRef);
                        }
                        catch (error) {
                            console.error('Error creating toast', error);
                            innerResolve(null);
                        }
                    });
                });
            });
            Promise.all(allPromises).then((refs) => resolve(refs));
        });
    }
    closeToast(id) {
        if (id) {
            const comp = this.hotToastComponentList.find((item) => item.toast.id === id);
            if (comp) {
                comp.close();
                this.cdr.markForCheck();
            }
        }
        else {
            this.hotToastComponentList.forEach((comp) => comp.close());
            this.cdr.markForCheck();
        }
    }
    beforeClosed(toast) {
        toast.visible = false;
        this.cdr.markForCheck();
    }
    afterClosed(closeToast) {
        const toastIndex = this.toasts.findIndex((t) => t.id === closeToast.id);
        if (toastIndex > -1) {
            this._onClosed.next(closeToast);
            this.toasts = this.toasts.filter((t) => t.id !== closeToast.id);
            this.toastRefs = this.toastRefs.filter((t) => t.getToast().id !== closeToast.id);
            this.cdr.markForCheck();
        }
    }
    toggleGroup(groupEvent) {
        const toastIndex = this.toastRefs.findIndex((t) => t.getToast().id === groupEvent.id);
        if (toastIndex > -1) {
            this._onGroupToggle.next(groupEvent);
            this.toastRefs[toastIndex].groupExpanded = groupEvent.event === 'expand';
            this.cdr.markForCheck();
        }
    }
    hasToast(id) {
        return this.toasts.findIndex((t) => t.id === id) > -1;
    }
    showAllToasts(show) {
        this.isShowingAllToasts = show;
    }
    ngOnDestroy() {
        if (this.defaultConfig().usePopover) {
            // We need the try/catch because the browser will throw if the
            // host or any of the parents are outside the DOM. Also note
            // the string access which is there for compatibility with Closure.
            try {
                this.host.nativeElement['hidePopover']();
            }
            catch (error) {
                if (isDevMode()) {
                    console.error('Error hiding popover');
                    console.error(error);
                }
            }
        }
    }
    getAfterClosed(toast) {
        return this.onClosed$.pipe(filter((v) => v.id === toast.id));
    }
    getAfterGroupToggled(toast) {
        return this.onGroupToggle$.pipe(filter((v) => v.id === toast.id));
    }
    getAfterGroupRefsAttached(toast) {
        return this.onGroupRefAttached$.pipe(filter((v) => v.id === toast.id));
    }
    updateToasts(toast, options) {
        this.toasts = this.toasts.map((t) => ({ ...t, ...(t.id === toast.id && { ...toast, ...options }) }));
        this.cdr.markForCheck();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastContainerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "21.2.9", type: HotToastContainerComponent, isStandalone: true, selector: "hot-toast-container", inputs: { defaultConfig: { classPropertyName: "defaultConfig", publicName: "defaultConfig", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.popover": "defaultConfig().usePopover ? \"manual\" : undefined", "class.hot-toast-container-overlay-popover": "defaultConfig().usePopover" } }, viewQueries: [{ propertyName: "hotToastComponentList", predicate: HotToastComponent, descendants: true }], ngImport: i0, template: "<div class=\"hot-toast-container-overlay\">\n  <div class=\"hot-toast-container-wrapper\">\n    <div>\n      @for (toast of toasts; track trackById($index, toast)) {\n        @if (toast.group?.parent) {\n        } @else {\n          <hot-toast-component\n            [toast]=\"toast\"\n            [offset]=\"calculateOffset(toast.id, toast.position)\"\n            [toastRef]=\"toastRefs[$index]\"\n            [toastsAfter]=\"(toast.autoClose ? toasts.length : getVisibleToasts(toast.position).length) - 1 - $index\"\n            [defaultConfig]=\"defaultConfig()\"\n            [isShowingAllToasts]=\"isShowingAllToasts\"\n            (showAllToasts)=\"showAllToasts($event)\"\n            (height)=\"updateHeight($event, toast)\"\n            (beforeClosed)=\"beforeClosed(toast)\"\n            (afterClosed)=\"afterClosed($event)\"\n            (toggleGroup)=\"toggleGroup($event)\"\n          />\n        }\n      }\n    </div>\n  </div>\n</div>\n", styles: [".hot-toast-container-overlay{position:fixed;z-index:var(--hot-toast-container-overlay-z-index, 9999);inset:0;pointer-events:none}.hot-toast-container-wrapper{position:relative;height:100%}\n"], dependencies: [{ kind: "component", type: HotToastComponent, selector: "hot-toast-component", inputs: ["toast", "offset", "defaultConfig", "toastRef", "toastsAfter", "isShowingAllToasts"], outputs: ["height", "beforeClosed", "afterClosed", "showAllToasts", "toggleGroup"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, preserveWhitespaces: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastContainerComponent, decorators: [{
            type: Component,
            args: [{ selector: 'hot-toast-container', changeDetection: ChangeDetectionStrategy.OnPush, imports: [HotToastComponent], host: {
                        '[attr.popover]': 'defaultConfig().usePopover ? "manual" : undefined',
                        '[class.hot-toast-container-overlay-popover]': 'defaultConfig().usePopover',
                    }, template: "<div class=\"hot-toast-container-overlay\">\n  <div class=\"hot-toast-container-wrapper\">\n    <div>\n      @for (toast of toasts; track trackById($index, toast)) {\n        @if (toast.group?.parent) {\n        } @else {\n          <hot-toast-component\n            [toast]=\"toast\"\n            [offset]=\"calculateOffset(toast.id, toast.position)\"\n            [toastRef]=\"toastRefs[$index]\"\n            [toastsAfter]=\"(toast.autoClose ? toasts.length : getVisibleToasts(toast.position).length) - 1 - $index\"\n            [defaultConfig]=\"defaultConfig()\"\n            [isShowingAllToasts]=\"isShowingAllToasts\"\n            (showAllToasts)=\"showAllToasts($event)\"\n            (height)=\"updateHeight($event, toast)\"\n            (beforeClosed)=\"beforeClosed(toast)\"\n            (afterClosed)=\"afterClosed($event)\"\n            (toggleGroup)=\"toggleGroup($event)\"\n          />\n        }\n      }\n    </div>\n  </div>\n</div>\n", styles: [".hot-toast-container-overlay{position:fixed;z-index:var(--hot-toast-container-overlay-z-index, 9999);inset:0;pointer-events:none}.hot-toast-container-wrapper{position:relative;height:100%}\n"] }]
        }], ctorParameters: () => [], propDecorators: { defaultConfig: [{ type: i0.Input, args: [{ isSignal: true, alias: "defaultConfig", required: false }] }], hotToastComponentList: [{
                type: ViewChildren,
                args: [HotToastComponent]
            }] } });

class ToastConfig {
    constructor() {
        /**
         * Sets the reverse order for hot-toast stacking
         *
         * @default false
         */
        this.reverseOrder = false;
        /**
         * Sets the number of toasts visible.
         * 0 will set no limit.
         * @default 5
         * @since 6.1.0
         */
        this.visibleToasts = 5;
        /**
         * Sets the type of stacking
         * @default "vertical"
         * @since 6.1.0
         */
        this.stacking = 'vertical';
        this.ariaLive = 'polite';
        this.role = 'status';
        this.position = 'top-center';
        this.autoClose = true;
        this.theme = 'toast';
        this.attributes = {};
        // key in ToastType
        this.info = { content: '' };
        this.success = { content: '' };
        this.error = { content: '' };
        this.loading = { content: '' };
        this.blank = { content: '' };
        this.warning = { content: '' };
    }
}
const isFunction = (valOrFunction) => typeof valOrFunction === 'function';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isAngularComponent = (arg) => {
    // `ɵcmp` is a component definition set by the Angular compiler.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return typeof arg === 'function' && !!arg.ɵcmp;
};
const resolveValueOrFunction = (valOrFunction, arg) => isAngularComponent(valOrFunction)
    ? valOrFunction
    : isFunction(valOrFunction)
        ? valOrFunction(arg)
        : valOrFunction;
class ToastPersistConfig {
    constructor() {
        /**
         *In which storage id vs. counts should be stored
         *
         * @type {('local' | 'session')}
         * @memberof ToastPersistConfig
         * @default 'local'
         */
        this.storage = 'local';
        /**
         *The key pattern to store object in storage. `${id}` in pattern is replaced with actual toast id.
         *
         * @type {('local' | 'session')}
         * @memberof ToastPersistConfig
         * @default 'ngxpert/hottoast-${id}'
         */
        this.key = 'ngxpert/hototast-${id}';
        /**
         *The number of toasts allowed to show.
         *
         * @memberof ToastPersistConfig
         * @default 1
         */
        this.count = 1;
        this.enabled = false;
    }
}

/**
 * Injection token to provide a custom container selector for the toast container. The value will be used as a selector to find the container element using `document.querySelector`.
 * @example
 * ```ts
 * import { HOT_TOAST_CONTAINER_TOKEN, provideHotToastConfig } from '@ngxpert/hot-toast';
 * import { bootstrapApplication } from '@angular/platform-browser';
 *
 * bootstrapApplication(AppComponent, {
 *   providers: [
 *     provideHotToastConfig(),
 *     {
 *       provide: HOT_TOAST_CONTAINER_TOKEN,
 *       useValue: '#toast-container',
 *     },
 *   ],
 * }).catch((err) => console.error(err));
 * ```
 *
 */
const HOT_TOAST_CONTAINER_TOKEN = new InjectionToken('HOT_TOAST_CONTAINER_TOKEN');
/** Injection token used to configure whether the toast should be rendered as a native popover element.
 *  * @example
 * ```ts
 * import { HOT_TOAST_USE_POPOVER_TOKEN, provideHotToastConfig } from '@ngxpert/hot-toast';
 * import { bootstrapApplication } from '@angular/platform-browser';
 *
 * bootstrapApplication(AppComponent, {
 *   providers: [
 *     provideHotToastConfig(),
 *     {
 *       provide: HOT_TOAST_USE_POPOVER_TOKEN,
 *       useValue: true,
 *     },
 *   ],
 * }).catch((err) => console.error(err));
 * ```
 *
 */
const HOT_TOAST_USE_POPOVER_TOKEN = new InjectionToken('HOT_TOAST_USE_POPOVER_TOKEN');

class HotToastService {
    static { this.nextId = 0; }
    constructor() {
        this._isInitialized = false;
        this._defaultGlobalConfig = new ToastConfig();
        this._defaultPersistConfig = new ToastPersistConfig();
        this._viewService = inject(ViewService);
        this._platformId = inject(PLATFORM_ID);
        this._globalConfig = inject(ToastConfig, { optional: true });
        this._container = inject(HOT_TOAST_CONTAINER_TOKEN, { optional: true });
        this._usePopover = inject(HOT_TOAST_USE_POPOVER_TOKEN, { optional: true });
        this._document = inject(DOCUMENT);
        if (this._globalConfig) {
            this._defaultGlobalConfig = {
                ...this._defaultGlobalConfig,
                ...this._globalConfig,
            };
        }
    }
    get defaultConfig() {
        return this._defaultGlobalConfig;
    }
    set defaultConfig(config) {
        this._defaultGlobalConfig = {
            ...this._defaultGlobalConfig,
            ...config,
        };
        if (this._componentRef) {
            this._componentRef.setInput('defaultConfig', this._defaultGlobalConfig);
        }
    }
    /**
     * Opens up an hot-toast without any pre-configurations
     *
     * @param message The message to show in the hot-toast.
     * @param [options] Additional configuration options for the hot-toast.
     * @param skipAttachToParent Only for internal usage. Setting this to true will not attach toast to it's parent.
     * @returns
     * @memberof HotToastService
     */
    show(message, options, skipAttachToParent) {
        const toast = this.createToast({
            message: message || this._defaultGlobalConfig.blank.content,
            type: options?.type ?? 'blank',
            options: {
                ...this._defaultGlobalConfig,
                ...options,
            },
            skipAttachToParent,
        });
        return toast;
    }
    /**
     * Opens up an hot-toast with pre-configurations for error state
     *
     * @param message The message to show in the hot-toast.
     * @param [options] Additional configuration options for the hot-toast.
     * @returns
     * @memberof HotToastService
     */
    error(message, options) {
        const toast = this.createToast({
            message: message || this._defaultGlobalConfig.error.content,
            type: 'error',
            options: {
                ...this._defaultGlobalConfig,
                ...this._defaultGlobalConfig?.error,
                ...options,
            },
        });
        return toast;
    }
    /**
     * Opens up an hot-toast with pre-configurations for success state
     *
     * @param message The message to show in the hot-toast.
     * @param [options] Additional configuration options for the hot-toast.
     * @returns
     * @memberof HotToastService
     */
    success(message, options) {
        const toast = this.createToast({
            message: message || this._defaultGlobalConfig.success.content,
            type: 'success',
            options: {
                ...this._defaultGlobalConfig,
                ...this._defaultGlobalConfig?.success,
                ...options,
            },
        });
        return toast;
    }
    /**
     * Opens up an hot-toast with pre-configurations for loading state
     *
     * @param message The message to show in the hot-toast.
     * @param [options] Additional configuration options for the hot-toast.
     * @returns
     * @memberof HotToastService
     */
    loading(message, options) {
        const toast = this.createToast({
            message: message || this._defaultGlobalConfig.loading.content,
            type: 'loading',
            options: {
                ...this._defaultGlobalConfig,
                ...this._defaultGlobalConfig?.loading,
                ...options,
            },
        });
        return toast;
    }
    /**
     * Opens up an hot-toast with pre-configurations for warning state
     *
     * @param message The message to show in the hot-toast.
     * @param [options] Additional configuration options for the hot-toast.
     * @returns
     * @memberof HotToastService
     */
    warning(message, options) {
        const toast = this.createToast({
            message: message || this._defaultGlobalConfig.warning.content,
            type: 'warning',
            options: {
                ...this._defaultGlobalConfig,
                ...this._defaultGlobalConfig?.warning,
                ...options,
            },
        });
        return toast;
    }
    /**
     * Opens up an hot-toast with pre-configurations for info state
     *
     * @param message The message to show in the hot-toast.
     * @param [options] Additional configuration options for the hot-toast.
     * @returns
     * @memberof HotToastService
     * @since 3.3.0
     */
    info(message, options) {
        const toast = this.createToast({
            message: message || this._defaultGlobalConfig.info.content,
            type: 'info',
            options: {
                ...this._defaultGlobalConfig,
                ...this._defaultGlobalConfig?.info,
                ...options,
            },
        });
        return toast;
    }
    /**
     *
     *  Opens up an hot-toast with pre-configurations for loading initially and then changes state based on messages
     *
     * @template T Type of observable
     * @param messages Messages for each state i.e. loading, success and error
     * @returns
     * @memberof HotToastService
     */
    observe(messages) {
        return (source) => {
            let toastRef;
            let start = 0;
            const loadingContent = messages.loading ?? this._defaultGlobalConfig.loading?.content;
            const successContent = messages.success ?? this._defaultGlobalConfig.success?.content;
            const errorContent = messages.error ?? this._defaultGlobalConfig.error?.content;
            return defer(() => {
                if (loadingContent) {
                    toastRef = this.createLoadingToast(loadingContent);
                    start = Date.now();
                }
                return source.pipe(tap({
                    ...(successContent && {
                        next: (val) => {
                            toastRef = this.createOrUpdateToast(messages, val, toastRef, 'success', start === 0 ? start : Date.now() - start);
                        },
                    }),
                    ...(errorContent && {
                        error: (e) => {
                            toastRef = this.createOrUpdateToast(messages, e, toastRef, 'error', start === 0 ? start : Date.now() - start);
                        },
                    }),
                }));
            });
        };
    }
    fromForm(control, options) {
        let toastRef;
        const handleStatus = async (status) => {
            const stateConfig = options[status];
            if (!stateConfig) {
                toastRef?.close();
                toastRef = undefined;
                return;
            }
            const { message, show, ...toastOptions } = stateConfig;
            const resolvedMessage = typeof message === 'function' ? message(control) : message;
            const statusDefaults = HOT_TOAST_FORM_STATUS_DEFAULTS[status];
            const mergedOptions = { ...statusDefaults, ...toastOptions };
            if (show) {
                const shouldShow = await show(control);
                if (!shouldShow) {
                    toastRef?.close();
                    toastRef = undefined;
                    return;
                }
                else {
                    if (toastRef) {
                        toastRef.updateMessage(resolvedMessage);
                        toastRef.updateToast(mergedOptions);
                    }
                    else {
                        toastRef = this.show(resolvedMessage, mergedOptions);
                        toastRef.afterClosed.subscribe(() => {
                            toastRef = undefined;
                        });
                    }
                }
            }
            // If show is not provided, proceed when form is dirty & touched
            else if (control.dirty && control.touched) {
                if (toastRef) {
                    toastRef.updateMessage(resolvedMessage);
                    toastRef.updateToast(mergedOptions);
                }
                else {
                    toastRef = this.show(resolvedMessage, mergedOptions);
                    toastRef.afterClosed.subscribe(() => {
                        toastRef = undefined;
                    });
                }
            }
        };
        handleStatus(control.status);
        const subscription = new Subscription();
        subscription.add(control.statusChanges.subscribe((status) => handleStatus(status)));
        return {
            close: () => {
                subscription.unsubscribe();
                toastRef?.close();
                toastRef = undefined;
            },
        };
    }
    /**
     * Closes the hot-toast
     *
     * @param [id] - ID of the toast
     * @since 3.0.1 - If ID is not provided, all toasts will be closed
     */
    close(id) {
        if (this._componentRef) {
            this._componentRef.ref.instance.closeToast(id);
        }
    }
    /**
     * Used for internal purpose only.
     * Creates a container component and attaches it to document.body.
     */
    init() {
        if (isPlatformServer(this._platformId)) {
            return;
        }
        const defaultUsePopover = this._usePopover ?? true;
        if (!('showPopover' in this._document.body)) {
            this._defaultGlobalConfig.usePopover = false;
        }
        else {
            this._defaultGlobalConfig.usePopover = this._defaultGlobalConfig.usePopover ?? defaultUsePopover;
        }
        if (this._container) {
            let containerElement = document.querySelector(this._container);
            if (!containerElement) {
                console.warn(`No container element found for selector: ${this._container}, using document.body instead as toast container.`);
                containerElement = document.body;
            }
            this._componentRef = this._viewService
                .createComponent(HotToastContainerComponent)
                .setInput('defaultConfig', this._defaultGlobalConfig)
                .appendTo(containerElement);
        }
        else {
            this._componentRef = this._viewService
                .createComponent(HotToastContainerComponent)
                .setInput('defaultConfig', this._defaultGlobalConfig)
                .appendTo(document.body);
        }
    }
    createOrUpdateToast(messages, val, toastRef, type, diff) {
        try {
            let content = null;
            let options = {};
            ({ content, options } = this.getContentAndOptions(type, messages[type] || (this._defaultGlobalConfig[type] ? this._defaultGlobalConfig[type].content : '')));
            content = resolveValueOrFunction(content, val);
            if (toastRef) {
                if (options.data) {
                    toastRef.data = options.data;
                }
                toastRef.updateMessage(content);
                const updatedOptions = {
                    type,
                    duration: diff + HOT_TOAST_DEFAULT_TIMEOUTS[type],
                    ...options,
                    ...(options.duration && { duration: diff + options.duration }),
                };
                toastRef.updateToast(updatedOptions);
            }
            else {
                this.createToast({ message: content, type, options });
            }
            return toastRef;
        }
        catch (error) {
            console.error(error);
        }
    }
    createToast({ message, type, options, observableMessages, skipAttachToParent, }) {
        if (!this._isInitialized) {
            this._isInitialized = true;
            this.init();
        }
        const id = options?.id ?? `toast-${HotToastService.nextId++}`;
        if (!this.isDuplicate(id) &&
            (!options.persist?.enabled || (options.persist?.enabled && this.handleStorageValue(id, options)))) {
            const toast = {
                ariaLive: options?.ariaLive ?? 'polite',
                createdAt: Date.now(),
                duration: options?.duration ?? HOT_TOAST_DEFAULT_TIMEOUTS[type],
                id,
                message,
                role: options?.role ?? 'status',
                type,
                visible: true,
                observableMessages: observableMessages ?? undefined,
                ...options,
            };
            return new HotToastRef(toast).appendTo(this._componentRef.ref.instance, skipAttachToParent);
        }
    }
    /**
     * Checks whether any toast with same id is present.
     *
     * @private
     * @param id - Toast ID
     */
    isDuplicate(id) {
        return this._componentRef.ref.instance.hasToast(id);
    }
    /**
     * Creates an entry in local or session storage with count ${defaultConfig.persist.count}, if not present.
     * If present in storage, reduces the count
     * and returns the count.
     * Count can not be less than 0.
     */
    handleStorageValue(id, options) {
        let count = 1;
        const persist = { ...this._defaultPersistConfig, ...options.persist };
        const storage = persist.storage === 'local' ? localStorage : sessionStorage;
        const key = persist.key.replace(/\${id}/g, id);
        let item = storage.getItem(key);
        if (item) {
            item = parseInt(item, 10);
            if (item > 0) {
                count = item - 1;
            }
            else {
                count = item;
            }
        }
        else {
            count = persist.count;
        }
        storage.setItem(key, count.toString());
        return count;
    }
    getContentAndOptions(toastType, message) {
        let content;
        let options = {
            ...this._defaultGlobalConfig,
            ...this._defaultGlobalConfig[toastType],
        };
        // typeof message === 'object' won't work, cz TemplateRef's type is object
        if (typeof message === 'string' || isTemplateRef(message) || isComponent(message)) {
            content = message;
        }
        else {
            let restOptions;
            ({ content, ...restOptions } = message);
            options = { ...options, ...restOptions };
        }
        return { content, options };
    }
    createLoadingToast(messages) {
        let content = null;
        let options = {};
        ({ content, options } = this.getContentAndOptions('loading', messages));
        return this.loading(content, options);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.9", ngImport: i0, type: HotToastService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: () => [] });

function provideHotToastConfig(config) {
    return makeEnvironmentProviders([{ provide: ToastConfig, useValue: config }]);
}

const DEFAULT_HOT_TOAST_HTTP_INTERCEPTOR_CONFIG = {
    ignoreStatuses: [],
    showForUnknownErrors: true,
};

const HOT_TOAST_HTTP_INTERCEPTOR_CONFIG = new InjectionToken('HOT_TOAST_HTTP_INTERCEPTOR_CONFIG');
function resolveErrorMessage(context) {
    const { error } = context;
    if (error instanceof HttpErrorResponse) {
        if (typeof error.error === 'string' && error.error.trim().length > 0) {
            return error.error;
        }
        if (error.message) {
            return error.message;
        }
    }
    if (error instanceof Error && error.message) {
        return error.message;
    }
    return 'Something went wrong';
}
function resolveToastMessage(config, context) {
    if (typeof config.errorMessage === 'function') {
        return config.errorMessage(context);
    }
    if (typeof config.errorMessage === 'string') {
        return config.errorMessage;
    }
    return resolveErrorMessage(context);
}
/**
 * Shows an error toast for failed HTTP calls. Register with
 * `provideHttpClient(withInterceptors([hotToastHttpInterceptor]))` (and your other HTTP features),
 * and optionally call {@link provideHotToastHttpInterceptor} for ignore rules and messages.
 */
const hotToastHttpInterceptor = (req, next) => {
    const platformId = inject(PLATFORM_ID);
    const toast = inject(HotToastService);
    const injectedConfig = inject(HOT_TOAST_HTTP_INTERCEPTOR_CONFIG, { optional: true });
    const config = {
        ...DEFAULT_HOT_TOAST_HTTP_INTERCEPTOR_CONFIG,
        ...injectedConfig,
    };
    if (config.skipRequest?.({ request: req })) {
        return next(req);
    }
    return next(req).pipe(catchError((error) => {
        if (isPlatformServer(platformId)) {
            return throwError(error);
        }
        const status = error instanceof HttpErrorResponse ? error.status : undefined;
        const url = error instanceof HttpErrorResponse ? error.url ?? undefined : undefined;
        const context = {
            error,
            request: req,
            status,
            url: url ?? req.url,
        };
        if (status !== undefined && config.ignoreStatuses?.includes(status)) {
            return throwError(error);
        }
        if (config.shouldIgnore?.(context)) {
            return throwError(error);
        }
        if (!(error instanceof HttpErrorResponse) && !config.showForUnknownErrors) {
            return throwError(error);
        }
        const message = resolveToastMessage(config, context);
        if (message) {
            toast.error(message, config.toastOptions);
        }
        return throwError(error);
    }));
};
/**
 * Optional configuration for {@link hotToastHttpInterceptor}. Add the interceptor to `HttpClient`, for example:
 *
 * ```ts
 * provideHttpClient(withFetch(), withInterceptors([hotToastHttpInterceptor])),
 * provideHotToastHttpInterceptor({ ignoreStatuses: [401] }),
 * ```
 */
function provideHotToastHttpInterceptor(config) {
    return makeEnvironmentProviders([
        {
            provide: HOT_TOAST_HTTP_INTERCEPTOR_CONFIG,
            useValue: { ...DEFAULT_HOT_TOAST_HTTP_INTERCEPTOR_CONFIG, ...config },
        },
    ]);
}

class HotToastBuilder {
    constructor(message, service) {
        this.message = message;
        this.service = service;
        this.options = {};
        this.groupChildren = [];
    }
    setOptions(options) {
        this.options = { ...this.options, ...options };
        return this;
    }
    addChild(child) {
        this.groupChildren.push(child);
        return this;
    }
    get afterGroupRefsAttached() {
        return this.toastRef?.afterGroupRefsAttached;
    }
    addChildrenToOptions() {
        if (this.groupChildren.length > 0) {
            const children = this.groupChildren.map((child) => ({
                options: {
                    message: child.message,
                    ...child.options,
                },
            }));
            this.options.group = {
                ...this.options.group,
                children,
            };
        }
    }
    // Create method that creates but doesn't show the toast. Call show() to show the toast.
    create(method = 'show') {
        this.addChildrenToOptions();
        this.toastRef = this.service[method](this.message, {
            ...this.options,
            ...{ visible: false },
        });
        return this.toastRef;
    }
    createToast(method) {
        this.addChildrenToOptions();
        this.toastRef = this.service[method](this.message, this.options);
        return this.toastRef;
    }
    show() {
        return this.createToast('show');
    }
    success() {
        return this.createToast('success');
    }
    error() {
        return this.createToast('error');
    }
    warning() {
        return this.createToast('warning');
    }
    info() {
        return this.createToast('info');
    }
    loading() {
        return this.createToast('loading');
    }
}

/*
 * Public API Surface of hot-toast
 */

/**
 * Generated bundle index. Do not edit.
 */

export { DEFAULT_HOT_TOAST_HTTP_INTERCEPTOR_CONFIG, HOT_TOAST_CONTAINER_TOKEN, HOT_TOAST_HTTP_INTERCEPTOR_CONFIG, HOT_TOAST_USE_POPOVER_TOKEN, HotToastBuilder, HotToastRef, HotToastService, ToastConfig, ToastPersistConfig, hotToastHttpInterceptor, provideHotToastConfig, provideHotToastHttpInterceptor, resolveValueOrFunction };
